src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,438 @@
+            {
+                _registeredHandlesLock.Acquire();
+                try
+                {
+                    if (_numUserWaits == WaitHandle.MaxWaitHandles - 1)
+                    {
+                        return false;
+                    }
+
+                    _registeredWaits[_numUserWaits] = handle;
+                    _waitHandles[_numUserWaits + 1] = handle.Handle;
+                    _numUserWaits++;
+
+                    handle.WaitThread = this;
+                }
+                finally
+                {
+                    _registeredHandlesLock.Release();
+                }
+
+                _changeHandlesEvent.Set();
+                return true;
+            }
+            
+            /// <summary>
+            /// Unregisters a wait handle.
+            /// </summary>
+            /// <param name="handle">The handle to unregister.</param>
+            /// <remarks>
+            /// As per CoreCLR's behavior, if the user passes in an invalid <see cref="WaitHandle"/>
+            /// into <see cref="RegisteredWaitHandle.Unregister(WaitHandle)"/>, then the unregistration of the wait handle is blocking.
+            /// Otherwise, the unregistration of the wait handle is queued on the wait thread.
+            /// </remarks>
+            public void UnregisterWait(RegisteredWaitHandle handle)
+            {
+                UnregisterWait(handle, handle.IsBlocking, true);
+            }
+
+            /// <summary>
+            /// Unregister a wait handle.
+            /// </summary>
+            /// <param name="handle">The wait handle to unregister.</param>
+            /// <param name="blocking">Should the unregistration block on the full unregistration and all pending callback completion.</param>
+            /// <param name="waitForRemoval">Should wait for the handle to be removed from the array before ending</param>
+            private void UnregisterWait(RegisteredWaitHandle handle, bool blocking, bool waitForRemoval)
+            {
+                bool pendingRemoval = false;
+                // TODO: Optimization: Try to unregister wait directly if it isn't being waited on.
+                _removesLock.Acquire();
+                try
+                {
+                    // If this handle is not already pending removal and hasn't already been removed
+                    if (Array.IndexOf(_registeredWaits, handle) != -1 && Array.IndexOf(_pendingRemoves, handle) == -1)
+                    {
+                        _pendingRemoves[_numPendingRemoves++] = handle;
+                        _changeHandlesEvent.Set(); // Tell the wait thread that there are changes pending.
+                        pendingRemoval = true;
+                    }
+                }
+                finally
+                {
+                    _removesLock.Release();
+                }
+
+                if (waitForRemoval && pendingRemoval)
```

Comment:
The UnregisterWait overload above may call this function with blocking: true and waitForRemoval: true. Suggestion:

```c#
if(blocking)
{
    // ...
}
else if(waitForRemoval && pendingRemoval)
{
    // ...
}
```